Added VOMongoRepository>>onlyUpdateIfCurrent: option

tinchodias opened the pull request:
When this option is enabled, the repository ensures the object's version is current before update.

If it's not current, a VOMongoSaveConflictError is signaled.

By default, this option is disabled.

tinchodias commented:
This would close #15

estebanlm commented:
test fails :disappointed:

tinchodias commented:
Now yes? it was a ridiculous random failure. In a new commit, I improved a bit the workaround in VOMongoTest>>tearDown. One day I will understand this annoying random failure!

tinchodias commented:
Ping! Now it's green, right?

zecke commented:
Can you point me to the random failure and your fix? Is it because a change has not been applied to the database yet?

zecke commented on mc/Voyage-Mongo-Core.package/VOMongoRepositoryResolver.class/instance/basicSave.id..st:
```diff
@@ -1,7 +1,13 @@
+			basicIsCurrent: anObject
+			id: idObject
+			ifFalse: [ VOMongoSaveConflictError signal ] ].
+
```
Sorry, that is not going to work in a reliable way. It is the classic time of check, time of action race condition.

You check.. ->
              <-  other image inserts
You compare your versions
You update ->

and you have a lost update...

https://github.com/zecke/osmo-smsc/blob/master/mc/ShortMessageCenter-Database.package/ShortMessageDatabase.class/instance/lockDestination.for..st for an example of findAndModify but db.collection.update is atomic as well.

So the update that is executed should be for the _id of the object and the version number. If not update was made, we have a conflict.

estebanlm commented:
As suggested by @zecke, this approach might work:

I see where we have a disconnect. I am thinking of single "documents" and not of graphs (e.g. when I save.. I never expect it to recursively save objects in my current application).

So if we take my.. >>#saveOne: mindset, I have added two examples in the ticket but let me repeat it here. update() and findAndModify() are atomic per document update.

```
db.collection.update({
	"where" : { '_id' : objectId, '#version' : expectedVersion },
       "update": {
		new..doc
	})
```

tinchodias commented:
@zecke about the random failure:

It is reproduced in `VORepositoryTest>>testSelectAllJustAfterRemove`. It's marked as an `expectedFailure`.

This is the code:

	testSelectAllJustAfterRemove
		"This test reproduces a bug"

		1 to: 30 do: [ :index |
			| pilot pilots |
		
			pilot := VOTestPilot new
				name: 'Esteban', index asString;
				yourself.

			self repository save: pilot.

			pilots := self repository selectAll: VOTestPilot.
			self deny: pilots isEmpty.

			self repository remove: pilot.
		
			pilots := self repository selectAll: VOTestPilot.
			self assert: pilots isEmpty.
		]

tinchodias commented:
Sorry, I have to correct my answer. In fact, the quoted test fails ALWAYS, not randomly. To see the random failure, you can just remove the last two lines from: 

	VOMongoTest>>
	tearDown
		"Reset the DB after each test"
		"self repository dropDatabase --> this works but it's slow and... leaks open Sockets!"

		self repository
			removeAll: VOTestPet;
			removeAll: VOTestPilot;
			flush.
		Smalltalk garbageCollect.

		800 milliseconds asDelay wait. "Wait to reduce the probability of a concurrency conflict"

Then run all Voyage tests and you will find some random yellow tests.
(Side note: Now I realized that `Smalltalk garbageCollect.` was enough as workaround)

As the method's comment states, `self repository dropDatabase` was a previous workaround... but it brought some other random failures when adding new tests.

tinchodias commented:
Thanks for the feedback on my changes. I understand the problem of this race condition and I'm trying to find how to translate your snippets to Voyage code. Right now I think this solution requires to modify Mongo's `UpdateOperation` to include the query.

zecke commented:
Why would dropDatabase leak a socket? That seems to be a bug by itself.

There are different phases a change will be applied:

   * committed your change to memory(!)
   * to the journal
   * to disk

this means for some time after the remove other clients might still see the "removed" data. This is by their design and if you want to have other guarantees one will need to:

 a.) use getLastError in mongo < 2.6 (i don't care too much about that)
 b.) use the writeConcern (see https://docs.mongodb.org/manual/reference/method/db.collection.remove/)

 I will have to add b.) in April/May for my customer.

zecke commented:
I don't think you need to update UpdateOperation. I am using an older version of Voyage and for me the VOMongoRepositoryResolver looks like this:

```
update: anObject serialized: jsonDictionary withId: aOID
	self pool withDatabase: [ :db |
		(self collectionAt: anObject class inDatabase: db) 
			replace: {'_id'->aOID} asDictionary
			with: (jsonDictionary
				at: '_id' put: aOID;
				yourself) ].

	^aOID
```

So.. you just need to add {#version ->currentVersion} to the to-be replaced dictionary and handle the failure...

tinchodias commented:
Sorry for the delay and thanks for the feedback, I'm coming back on the code later today!

tinchodias commented:
I started but didn't fnish to reimplement the "update if current" feature via `VOMongoRepositoryResolver>>update:serialized:withId:` using  `{#version ->currentVersion}`

secondary topic:
WriteConcern are interesting. 
   1. This should be implemented at Mongo level, shouldn't it?
   2. Should we change repository save operations to acknowledge write operations by default, or as an option?

zecke commented:
> On 22 Apr 2016, at 06:00, Martín Dias <notifications@github.com> wrote:
> 
> I started but didn't fnish to reimplement the "update if current" feature via VOMongoRepositoryResolver>>update:serialized:withId: using {#version ->currentVersion}
> 
> secondary topic:
> WriteConcern are interesting.
> 
> 	• This should be implemented at Mongo level, shouldn't it?
> 	• Should we change repository save operations to acknowledge write operations by default, or as an option?


WriteConcern should probably be in a separate issue. I will need a solution for it in the short term as well. There are some constraints.

* It only works with MongoDB >= v2.6.
* The place it should be. For many other systems it can be set per database or per query.

In the use-case of tests. For something like Write then Read you would like to set the WriteConcern for the write. For >>#removeAll: if it is possible you want to set it as well.

I have an application where I keep short-term state of a "local" system (consisting out of multiple images) and "global" state. For modifications of the global state I always want to have the write be committed, for the local one I care less about it. I use the same VOMongoRepository to query the "local" and the "global" state.

I was planning to put that info into the VOMongoContainer. But what is that info? The WriteConcern can be very specific to the number of nodes, or be configured to only include the "majority". E.g. if I write-down the number of nodes, how would I make that configurable (somebody might run the app with three nodes, or 17). In terms of your tests you probably only want it to be committed to the journal...

A lot of options, and we should understand some more usescases first to see where it should be put. For your failing tests, maybe we are lucky and older versions ignore the writeConcern and you can just add it. Or introduce a call to getLastError() (on the same MongoDatabase connection) after the remove?

tinchodias commented:
New commit ready, with a completely different approach based on your reviews. No idea why my test in VOMultipleImageTest hangs... tomorrow I'll take another look.

The commit's diff is noisy because of old code in the branch that I cleaned up. So, following I paste the key method where everything is solved. Maybe I should recreate the changes from scratch to keep a clean project history... I see the branch has conflict too.

	update: anObject serialized: jsonDictionary withId: aOID
		| currentVersion lastError |
		currentVersion := self cacheVersionAt: aOID.

		self pool withDatabase: [ :db |
			(self collectionAt: anObject class inDatabase: db) 
				replace: {'_id'->aOID. '#version'->currentVersion} asDictionary
				with: (jsonDictionary
					at: '_id' put: aOID;
					yourself).
				
			self flag: #todo. "Move to Mongo"
			"See more: https://docs.mongodb.org/manual/reference/command/getLastError/"
			lastError := db command: {'getLastError'->1} asDictionary.
			(lastError at: 'n') isZero ifTrue: [ VOMongoSaveConflictError signal ] ].

		^aOID


My path / how I ended with this solution:

I continue learning of mongodb and Voyage+Mongo implementation. First, I tried to use the [WriteResult](https://docs.mongodb.org/manual/reference/method/WriteResult/#WriteResult) of the update operation, but then realized it's not accessible when using the [wire protocol](https://docs.mongodb.org/manual/reference/mongodb-wire-protocol) that our Mongo implementation uses to update. Then, I ended using [getLastError](https://docs.mongodb.org/manual/reference/command/getLastError/) as @zecke mentioned at the end of the message about WriteConcerns.

Ah, also I removed the `onlyUpdateIdCurrent` option I added to `VoMongoRepository` earlier in this branch... it looked not necessary. Not sure about that. 

Thanks in advance for any feedback.

tinchodias commented:
Ah, apparently the test suffers the [infinite loop issue](https://github.com/estebanlm/voyage/issues/5)

tinchodias commented:
And... another side note. I must check what is wrong in my set up of [merge driver](https://github.com/ThierryGoubier/GitFileTree-MergeDriver). I get many git file conflicts that I guess the driver should automatically tackle!

noha commented:
What is the the problem with the wire protocol our mongo driver uses? Can you elaborate? Doing the getLastError every time is IMO no good idea. It just slows everything down. What keeps us from using the results on an update/insert directly?

zecke commented:
> On 26 Apr 2016, at 09:45, Norbert Hartl <notifications@github.com> wrote:
> 
> What is the the problem with the wire protocol our mongo driver uses? Can you elaborate? Doing the getLastError every time is IMO no good idea. It just slows everything down. What keeps us from using the results on an update/insert directly?

https://docs.mongodb.org/manual/reference/mongodb-wire-protocol/

```
Starting with MongoDB 2.6 and maxWireVersion 3, MongoDB drivers use the database commands insert, update, and delete instead of OP_INSERT, OP_UPDATE, and OP_DELETE for acknowledged writes. Most drivers continue to use opcodes for unacknowledged writes.
```

so we would either require MongoDB 2.6 or check the version and then do insert/update/delete through the command (adding a bit of overhead but allows us to do writeConcern handling).

The big question is.. how to control the consistency/speed one wants to have. E.g. for a "write" then "read" benchmark I run.. I sometimes see failing reads (as the data has not been committed yet) and would like to be able to control the write/update behavior.

noha commented:
I think there should a strategy object that implements the desired behaviour. One strategy would be FireAndForget to do the old default behaviour. Other strategies could implement the legacy way of checking by doing OP_...modify.. requests followed by getLastError. And a up-to-date strategy that uses commands and write concern. IMHO the strategy can be a configuration thing. Maybe there is an easy way to detect the wire protocol version than the strategy could be selected automatically on establish time of a connection

zecke commented:
> On 26 Apr 2016, at 10:37, Norbert Hartl <notifications@github.com> wrote:
> 
> I think there should a strategy object that implements the desired behaviour. One strategy would be FireAndForget to do the old default behaviour. Other strategies could implement the legacy way of checking by doing OP_...modify.. requests followed by getLastError. And a up-to-date strategy that uses commands and write concern. IMHO the strategy can be a configuration thing. Maybe there is an easy way to detect the wire protocol version than the strategy could be selected automatically on establish time of a connection

In Voyage or MongoTalk? Should MongoTalk expose all the different write variants first and then have Voyage pick it as it suits?

estebanlm commented:
> On 26 Apr 2016, at 13:37, Holger Freyther <notifications@github.com> wrote:
> 
> 
> > On 26 Apr 2016, at 10:37, Norbert Hartl <notifications@github.com> wrote:
> > 
> > I think there should a strategy object that implements the desired behaviour. One strategy would be FireAndForget to do the old default behaviour. Other strategies could implement the legacy way of checking by doing OP_...modify.. requests followed by getLastError. And a up-to-date strategy that uses commands and write concern. IMHO the strategy can be a configuration thing. Maybe there is an easy way to detect the wire protocol version than the strategy could be selected automatically on establish time of a connection
> 
> In Voyage or MongoTalk? Should MongoTalk expose all the different write variants first and then have Voyage pick it as it suits?

it should be in voyage: mongotalk is just a driver that maps directly… things like “version” (and then write strategies) are voyage specific. 

Esteban 

> 
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub <https://github.com/estebanlm/voyage/pull/21#issuecomment-214712990>

zecke commented:
> On 26 Apr 2016, at 13:43, Esteban Lorenzano <notifications@github.com> wrote:
> 
> 
> 
> 
> it should be in voyage: mongotalk is just a driver that maps directly… things like “version” (and then write strategies) are voyage specific. 

Yes, but MongoDB has two different protocol to Update/Delete/Insert. The second one is through their generic $cmd interface. The question is while MongoTalk already exposes a way to run a command, maybe exposing the three new commands makes sense?

tinchodias commented:
@noha the wire protocol doc says in update section that "There is no response to an OP_UPDATE message", while for example, the query operation "The database will respond to an OP_QUERY message with an OP_REPLY message". 

The idea of the strategies sounds good to implement alternative ways of performing the Mongo operations. As @estebanlm says, this can be at Voyage level. Also, a "command" strategy can use the Mongo command interface as @zecke proposes. To do that, I guess the "comamnd" strategy can directly use `MongoDatabase>>command:`, as I did for ´getLastError` in the snippet pasted above. 

BTW, now I see, it can be undesirable to have the performance penalty of getLastError in every case, then it should be optional. I'd be cool to benchmark the different strategies.

I'm trying to figure out what'd be this strategy hierarchy. Please review my sketch of classes and APIs:

**VOMongoInterfaceStrategy** abstract superclass
--> #isAvailable [testing]
--> #isFireAndForget, #fireAndForget: [accessing]
--> #update:... [operations]
--> #update:...writeConcern: [operations]
--> ... similar for insert, remove, and other operations

**VOMongoWireProtocolStrategy** subclass
--> #update:... (if not #isFireAndForget, acknowledge by reading **getLastError**)
--> #update:...writeConcern: (**signal error**)

**VOMongoRunCommandStrategy** subclass
--> #isAvailable (checks if Mongo version >=2.6)
--> #update:... (if not #isFireAndForget, acknowledge by reading **WriteResult**)
--> #update:...writeConcern: (**implement it**)

**VOMongoRepository** 
--> #useWireProtocol, etc. (new API to set interface strategy)
--> #update:...writeConcern: (new "extended" operations, which just forward, thus signal error if not supported by current strategy)

noha commented:
@tinchodias I do not understand the purpose of these classes. And it looks complex. All calls are rooted in the Mongo class. We care about insert:/update:/delete: This are the commands that change state and with the OP_... calls didn't return a result. So I would have a MongoWriteStrategy that can be set to the Mongo object. The first subclass would be MongoFireAndForgetWriteStrategy. It would get the current implementation of insert/update/delete in Mongo of these methods (using OP_...). And mongo delegates all these calls to the strategy. With the fire and forget being the default strategy if not set. From that point on the code should behave exactly like before. Then we can have MongoFireAndCheckWriteStrategy that uses the same like FireAndForget but after the write calls getLastError. On error it throws an exception. Finally the MongoCommandWriteStrategy would use the command call instead of the OP_ calls and gets a result. The result could also throw the same set of exceptions like FireAndCheck. The MongoCommandWriteStrategy would have extra accessors to set the writeConcern parameters. Furthermore we could have methods that set the writeConcern on the call while the FireAndForget and FireAndCheck would not impement and throw a configuration error

noha commented:
I forget to mention that I do not agree with Esteban. These things need to be in Mongo code because these are modes how to driver works. The basic abstraction voyage needs is in the insert:/update:/delete: calls. In voyage we just need a set of methods that allow us to take influence how the mongo objects are created. This would be needed to set the write strategy.

zecke commented:
> On 27 Apr 2016, at 11:00, Norbert Hartl <notifications@github.com> wrote:
> 
> I forget to mention that I do not agree with Esteban. These things need to be in Mongo code because these are modes how to driver works. The basic abstraction voyage needs is in the insert:/update:/delete: calls. In voyage we just need a set of methods that allow us to take influence how the mongo objects are created. This would be needed to set the write strategy.

I don't think it is one or the other. For my current application I have different objects with different consistency requirements.

a.) I create different VOMongoRepositories (increasing the amount of DB connections)
b.) There are two configure items. Something like minimal consistency set for the DB and then stronger one for the VOMongoCollection Magritte type?

noha commented:
I'm pretty sure it needs to be something that is implemented in the mongo code. I want to use the same thing if I'm only using the Mongo code because it is purely mongo related. How much of this is exposed to client code is a good question. I said the udate:/... method could have an addtional parameter for something like writeConcern. But what could be the model to integrate something like write concerns into an object serializer which voyage is. There needs to be a more general model than the write concern for mongo if voyage want to be independent of the backend.

tinchodias commented:
Hi, I'm back on this this issue. 

I didn't consider the backend-independence concern in Voyage... 
but I see there is already Mongo-specific API such as `selectMany:where:limit:offset:` and `dropDatabase`. They're implemented in VOMongoRepository but not in VORepository. There is also the authentication API. So, it woudln't be inconsistent to have write concern in VOMongoRepository. Anyway, this issue is about "update if current" :) 

Thanks for the feedback on my proposed hierarchy. it was a fast sketch, to continue discussion on something concrete... but @noha, I think we can implement the one you suggested. However, I still agree with @estebanlm. I don't see why Mongo-level is affected by this requirements: It already provides what's needed.

@zecke, I didn't understand if a.) and b.) imply that there is a 3rd position between @noha and @estebanlm. Could you explain more, please?

BTW, github web ui says that "@estebanlm added a commit that referenced this pull request" but I couldn't find where is the reference.

noha commented:
We need to figure what is really needed. A protocol of selectMany:where:limit:offset does not sound very mongo specific to me. Neither does dropDatabase. I talked to Esteban and a lot of the stuff that has mongo in its name is generic. So what we need to do is to pull out the generic parts from Voyage-Mongo into Voyage-Core. 
What do you mean by "I don't see why mongo-level is affected.."? We were talking about using return codes in mongo. Voyage can work without but it would be a feature to enable error handling. The enabled feature on the voyage side will change the implementation of the mongo driver.

tinchodias commented:
With "I don't see why mongo-level is affected.." I tried to say that while `Wire Protocol` update/insert operations need special support at MongoTalk-level (it'd be bad to write the bytes to the stream directly from Voyage), instead the update/insert operations via command are not that bad just using MongoDatabase>>command:. Concretely, this is how a "fire and forget" update via command looks more or less like:

```
commandFireAndForgetUpdate: anObject serialized: jsonDictionary withId: aOID 
	self pool withDatabase: [ :db |
		db command: {
			'update'->self collectionNameFor: anObject class.
			'updates'->{
				'q'->{'_id'->aOID} asDictionary.
				'u'->jsonDictionary at: '_id' put: aOID; yourself.
				} asDictionary.
			} asDictionary ].
	^aOID
```

versus how it currently looks via wire protocol:

```
wireProtocolFireAndForgetUpdate: anObject serialized: jsonDictionary withId: aOID
	self pool withDatabase: [ :db |
		(self collectionAt: anObject class inDatabase: db) 
			replace: {'_id'->aOID} asDictionary
			with: (jsonDictionary at: '_id' put: aOID; yourself) ].
	^aOID
```

tinchodias commented:
But ok, my affirmation was debatable. I recognize that the knowledge about those 'q' and 'u' parameters are low-level and fit better in MongoTalk. 

```
wireProtocolFireAndForgetUpdate: anObject serialized: jsonDictionary withId: aOID
	self pool withDatabase: [ :db |
		(self collectionAt: anObject class inDatabase: db) 
			replace: {'_id'->aOID} asDictionary
			with: (jsonDictionary at: '_id' put: aOID; yourself) ].
	^aOID
```

So, I have just pushed these changes to my branch https://github.com/tinchodias/voyage/tree/update_if_current

which requires this MongoTalk's version:

> Name: Mongo-Core-MartinDias.52
> Author: MartinDias
> Time: 9 May 2016, 2:36:26.2765 am
> UUID: 9ef50f3e-2e1f-4ac7-b4e2-843e0943057c
> Ancestors: Mongo-Core-NorbertHartl.50
> 
> Added experimental #commandUpdate:... API which is analog to #update:... API but using "Mongo command" instead of "Mongo Wire Protocol".

tinchodias commented:
Still have to check tests...

tinchodias commented:
A small update. I expect to soon come back to this branch, and get ready the new write strategies (WireProtocol and Command). Each strategy optionally acknowledges the write operation in it's own way (getLastError vs. WriteResult). It was working quite well in the branch... However, I wasn't sure how to nicely test the new 4 combinations ({WireProtocol,Command} x {FireAndForget, FireAndCheck}).

zecke commented:
> On 12 May 2016, at 09:04, Martín Dias <notifications@github.com> wrote:
> 
> A small update. I expect to soon come back to this branch, and get ready the new write strategies (WireProtocol and Command). Each strategy optionally acknowledges the write operation in it's own way (getLastError vs. WriteResult). It was working quite well in the branch... However, I wasn't sure how to nicely test the new 4 combinations ({WireProtocol,Command} x {FireAndForget, FireAndCheck}).

Great. If I want "ACKed" writes. Do you check the Mongo protocol version supported by the database? Testing will be difficult. You could have the DB on a mountpoint, start mongo and then fill the partition, then write.. it should fail.. but other than that, don't bother.

#1 We don't want to test mongodb
#2 The version of mongodb we test might not be the deployed one

noha commented:
I don't think it is necessary to test protocol version. It is heavy to do and after all that is an configuration issue. If someone is not capable of configuring the database and the driver accordingly he gets what he deserves ;)

zecke commented:
> On 13 May 2016, at 15:11, Norbert Hartl <notifications@github.com> wrote:
> 
> I don't think it is necessary to test protocol version. It is heavy to do and after all that is an configuration issue. If someone is not capable of configuring the database and the driver accordingly he gets what he deserves ;)


recursive exceptions until the VM crashes? I wanted to say that most headers contain the wire version.. but that is not the case. So yes, let's ignore it.

tinchodias commented:
Prototype should work by loading [my latest commit](https://github.com/estebanlm/voyage/pull/21/commits/dd64b945c9eb0a3c4043b413b3297a489476ed77). 

These are the configuration methods:
```
VOMongoRepository
  updateIfCurrent: true|false
  useCommandWriteStrategy.
  useWireProtocolWriteStrategy.
```

Since automated testing for this issue has been a headhache, I wrote this script to manually test the write strategies:

```
"--- SCRIPT 1 (IMAGE A): save initial object ---"

repository := VOMongoRepository database: 'Voyage-Tests'.	
dog := VOTestDog new name: 'initial'; yourself.
repository save: dog.
dogId := repository idOf: dog. "copy to clipboard  this dogId for next scripts"


"--- SCRIPT 2 (IMAGE B, i.e. IN OTHER IMAGE): change dog name to bar ---"

dogId := xxxxx.
repository := VOMongoRepository database: 'Voyage-Tests'.
dog := repository selectOne: VOTestDog where: (Dictionary with: '_id' -> dogId).
dog name: 'bar'. 
repository save: dog.


"--- SCRIPT 3 (IMAGE A, i.e. in the original one) ---"

dogId := xxxxx.
repository := VOMongoRepository database: 'Voyage-Tests'.	

"Set up a specific combination"
repository 
	updateIfCurrent: true;
	useCommandWriteStrategy.
"	useWireProtocolWriteStrategy."

"The dog is locally outdated respect to the db, so this should raise VOMongoSaveConflictError"
dog name: 'foo'. 
repository save: dog.
	
"The local update was not done, but updated object can be queried"
(repository selectOne: VOTestDog where: (Dictionary with: '_id' -> dogId)) name.
```

(I'm back into this issue after investing time last week into #28 :) )

tinchodias commented:
Im sorry, script 3 has a mistake: it shouldn't create a new repository. Also dogId is not necessary: that code should be run in the same Pharo workspace as script 1.

tinchodias commented:
About a question I didn't answer: no, the prototype doesn't check the Mongo protocol version supported by the database at the moment.

zecke commented:
> On 18 May 2016, at 07:15, Martín Dias <notifications@github.com> wrote:
> 
> Prototype should work by loading my latest commit.
> 
> These are the configuration methods:
> 
> VOMongoRepository
>   updateIfCurrent: true|false
>   useCommandWriteStrategy.
>   useWireProtocolWriteStrategy.

Great, one question. I wanted to use >>#updateIfCurrent: at one specific place in my code. Setting it for the entire repository seems to be a bit drastic and potentially causing more harm (updates failing when I actually don't care what the previous state was). Is it still possible to use it at one specific place only?

holger

tinchodias commented:
Good point. We can always add a `VOMongoRepository>>save:updateIfCurrent:`, but... if we want to apply this requirement to several options at the same time, we could have something like:

```
repository restoreAfter: [
   repository 
      useCommandWriteStrategy;
      updateIfCurrent: true;
      readPreference: blabla;
      writeConcern: bleble.

   repository save: specialObject
]
```

(inspired from Athens' restoreAfter:)

too complex?

this should take care of threads but I think it isn't a problem.

tinchodias commented:
I mean: 

```
restoreAfter: aBlock
	self mutex critical: [
		aBlock value: self copy ]
```

and:

```
mutex
	^mutex ifNil: [mutex := Semaphore forMutualExclusion]
```

In VORepository. Subclasses can reimplement `copy` to ensure the options are preserved after.


again... too complex?

tinchodias commented:
Or... as a more general alternative... it may be good idea to use a 'save options' object. In this case, a new `VORepository>>save:options:` method would receive a 'save options' argument that overrides the current repository's 'save options' (to set up through a new `VORepository>>defaultSaveOptions:`).

According to [mongodb update's documentation](https://docs.mongodb.com/manual/reference/command/update/#dbcmd.update), the operation can have the following options:

```
   ordered: <boolean>,
   writeConcern: { <write concern> },
   bypassDocumentValidation: <boolean>
```

This number of save options can justify this approach, but Voyage doesn't have them yet, so we can wait until we actually implement them and meanwhile use something simpler, or not.

(sorry if I'm thinking out loud too much!)

zecke commented:
> On 18 May 2016, at 22:09, Martín Dias <notifications@github.com> wrote:
> 
> Good point. We can always add a VOMongoRepository>>save:updateIfCurrent:, but... if we want to apply this requirement to several options at the same time, we could have something like:
> 
> repository restoreAfter: [
>    repository 
>       useCommandWriteStrategy;
>       updateIfCurrent: true;
>       readPreference: blabla;
>       writeConcern: bleble.
> 
>    repository save: specialObject
> ]
> 
> (inspired from Athens' restoreAfter:)
> 
> too complex?

too complex. Needs to handle concurrency correctly (to only apply to the current process) and even then still the above facade.

To provide the context again. Let's assume you track a TCP connection and store state in MongoDB. And to load-balance you run multiple images at the same time. In my application I track state of connections and some I don't mind if they overwrite the previous one and with others I do. As there are multiple workers at the same time, the updateIfCurrent setting "can" not be global.


holger

tinchodias commented:
Resuming the old discussion: Do you think it's fine to have: 
- #updateIfCurrent: global/default value and 
- #saveIfCurrent: and #save: id:updateIfCurrent: to override the global value for a specific operation

I implemented them in a24a938, and test passes. 
I still want to rethink the VOMongoWriteStrategy... I don't like it.

tinchodias commented:
Well, IMO this branch is finally ready for review. Main points to consider in this branch:

# Summary

VOMongoRepository has now:
- #saveIfCurrent: and #save: id:updateIfCurrent:, when save-if-current enabled, update the object and check if the object was still current in the database; else, signal a VOMongoSaveConflictError. Note: no change on inserts, only on update.
- #updateIfCurrent:, specifies the default option for #save: operations (default: false)
- #useWireProtocolToWrite:, true to use the old-style implmentation for #save: operations; false to use the [new WriteConcerns](https://docs.mongodb.com/manual/release-notes/2.6-compatibility/#write-method-acknowledgements), introduced in mongodb 2.6. (default: true)

As I asked in another PR, please tell me if you prefer that I rebase and create concise commits before the review. However, this summary might be enough to discuss the changes.

estebanlm commented:
so, what with this? integrate?

zecke commented:
Rebase/Squash and look at the condensed change? Need to check if the test verifies an exception being signaled.

tinchodias commented:
@estebanlm I think first it's better to finish splitting #45; the VOMongoExecuteStrategy stuff (#47) was one of the parts. 
@zecke, you can take a look on VOUpdateIfCurrentTest if you want to see this feature in action. I can split these changes too, ok.

tinchodias commented:
So next steps are:
- split changes to repository tests that are orthogonal to this PR
- re-check I don't forget anything in this PR
- condense PR (i.e. create a new one this is condensed in 1 commit)

tinchodias commented:
I create PR #62 which replaces this old branch with a clean one.
